Shell changes towards enabling an alternative shell (backward compatible with bash)

happysalada opened the pull request:
###### Motivation for this change

[Oil shell](https://www.oilshell.org) is a new shell that aims to retain compatibility with bash but brings in new feature that simplify shell scripting. Using it in nix would make shell scripting more accessible to many people (IMO). The drawback being that osh has not reached 1.0 yet and that there might still be some bugs.

This PR starts with a mix of shellcheck fixes, the ultimate goal being to be able to change the default shell shipped with stdenv.

Current state: Only shell fixes have been added to this PR at the moment, so stdenv should build fine. I am currently working on making the shell change work. The current error is related to a shell reference being included in the stdenv-stage3. At the moment stdenv-stage3 explicitely defines an `allowedRequisites` for bash.

@zimbatm I'm wondering who else might be interested by this. Specifically who we should make sure to expose early enough to know if this will be a no-go or not. 

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [X] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

happysalada committed:
builder.sh: shellcheck 2086

https://github.com/koalaman/shellcheck/wiki/SC2086

happysalada committed:
builder.sh: shellcheck 2129

https://github.com/koalaman/shellcheck/wiki/SC2129

happysalada committed:
builder.sh: check if variable is non null

This is not the "correct" way to check if a variable is non null in
bash. There is already an instance of the "right" way to do it in
setup.sh. Bash is "generous" enough to accept the original input though.
I couldn't find the relevant shellcheck.

happysalada committed:
setup.sh: shellcheck 2086

https://github.com/koalaman/shellcheck/wiki/SC2086

happysalada committed:
setup.sh: shellcheck 2206 + 2207

https://github.com/koalaman/shellcheck/wiki/SC2206
https://github.com/koalaman/shellcheck/wiki/SC2207

admittedly this one is a lot less pretty

happysalada committed:
setup.sh: arithmetic fixes

this one is a little more controversial
see https://github.com/oilshell/oil/issues/864
for more information

happysalada commented:
Here is the last error I got, after explicitely adding oil to the set of allowedRequisites
```
error: output '/nix/store/v0vs6aw6mpxcy2lsdnz3k5q0a8kr4jyf-bootstrap-stage3-stdenv-darwin' is not allowed to refer to the following paths:
         /nix/store/aqj0kzpan8z7y39qsy7mfawppvik50gm-ncurses-6.2
         /nix/store/hv7s7fgnlqrprld2xpp3brj24am0j2vb-readline-6.3p08
```

happysalada commented:
Next:
I'm trying to separate oil into an interactive version and a non-interactive one. (including readline or not, since it seems readline is used for interactivity).
I'm adding ncurses6 to the list of allowed requisites of stdenv  stage3 since oil depends on it.

happysalada commented:
adding the dependencies to allowedRequisites lead me to the next failure
```
❯ nix log /nix/store/kwb08d6ikz5hqkn2l83j2ap91953mfp0-gawk-5.1.0.drv
+ source '/nix/store/rl2wxq8qakzgxh1hslggk6vn7xgq5g5w-bootstrap-stage3-stdenv-darwin/setup'
+ export SHELL='/nix/store/yynw6wnds9q3kzbald33cia3prmd3qmy-bash-oil/bin/bash'
+ initialPath='/nix/store/6jysbyhc43sjvfiyh1bpvi1n3zbz212r-bootstrap-tools'
+ defaultNativeBuildInputs='/nix/store/4q88vhj678qfp5zz72k2vkmxfh6qn0h5-xz-5.2.5-bin /nix/store/mjjy30kxz7
+ defaultBuildInputs='/nix/store/h7ryjlg66nzh253m98zpcf6f8fjf09nc-swift-corefoundation /nix/store/yynw6wnd
+ export NIX_ENFORCE_NO_NATIVE=1
+ export NIX_ENFORCE_PURITY=1
+ export NIX_IGNORE_LD_THROUGH_GCC=1
+ unset SDKROOT
+ export gl_cv_func_getcwd_abort_bug=no
+ stripAllFlags=' '
+ export PATH='/nix/store/yynw6wnds9q3kzbald33cia3prmd3qmy-bash-oil/bin:/path-not-set'
+ export PATH_LOCALE='/nix/store/a37a66i5xl1ad3cdxsyjapcwhc0wa83x-adv_cmds-119-locale/share/locale'
+ export NIX_DONT_SET_RPATH_FOR_BUILD=1
+ export NIX_DONT_SET_RPATH=1
+ export NIX_NO_SELF_RPATH=1
+ export MACOSX_DEPLOYMENT_TARGET=10.12
+ set -eu
+ set -o pipefail
+ (( "${NIX_DEBUG:-0}" >= 6 ))
+ ':' 'out info man'
+ trap exitHandler EXIT
+ export SOURCE_DATE_EPOCH=
+ ':' 315532800
+ shopt -s nullglob
+ PATH=''
+ HOST_PATH=''
+ '[' '/nix/store/6jysbyhc43sjvfiyh1bpvi1n3zbz212r-bootstrap-tools' '=' '/' ']'
+ addToSearchPath PATH '/nix/store/6jysbyhc43sjvfiyh1bpvi1n3zbz212r-bootstrap-tools/bin'
+ addToSearchPathWithCustomDelimiter ':' PATH '/nix/store/6jysbyhc43sjvfiyh1bpvi1n3zbz212r-bootstrap-tools
+ local delimiter=':'
+ local varName=PATH
+ local dir='/nix/store/6jysbyhc43sjvfiyh1bpvi1n3zbz212r-bootstrap-tools/bin'
+ [[ -d "$dir" && "${!varName:+${delimiter}${!varName}${delimiter}}" \ ...
+ export PATH='/nix/store/6jysbyhc43sjvfiyh1bpvi1n3zbz212r-bootstrap-tools/bin'
+ '[' -z '' ']'
+ addToSearchPath HOST_PATH '/nix/store/6jysbyhc43sjvfiyh1bpvi1n3zbz212r-bootstrap-tools/bin'
+ addToSearchPathWithCustomDelimiter ':' HOST_PATH '/nix/store/6jysbyhc43sjvfiyh1bpvi1n3zbz212r-bootstrap-
+ local delimiter=':'
+ local varName=HOST_PATH
+ local dir='/nix/store/6jysbyhc43sjvfiyh1bpvi1n3zbz212r-bootstrap-tools/bin'
+ [[ -d "$dir" && "${!varName:+${delimiter}${!varName}${delimiter}}" \ ...
+ export HOST_PATH='/nix/store/6jysbyhc43sjvfiyh1bpvi1n3zbz212r-bootstrap-tools/bin'
+ unset i
+ (( "${NIX_DEBUG:-0}" >= 1 ))
+ '[' -z '/nix/store/yynw6wnds9q3kzbald33cia3prmd3qmy-bash-oil/bin/bash' ']'
+ BASH='/nix/store/yynw6wnds9q3kzbald33cia3prmd3qmy-bash-oil/bin/bash'
+ export CONFIG_SHELL='/nix/store/yynw6wnds9q3kzbald33cia3prmd3qmy-bash-oil/bin/bash'
+ '[' -z '' ']'
+ export shell='/nix/store/yynw6wnds9q3kzbald33cia3prmd3qmy-bash-oil/bin/bash'
+ runHook preHook
...skipping...
+ local hook=
+ _eval '_callImplicitHook 0 failureHook'
+ eval '_callImplicitHook 0 failureHook'
+ _callImplicitHook 0 failureHook
+ local def=0
+ local hookName=failureHook
+ declare -F failureHook
+ type -p failureHook
type: 'failureHook' not found
+ '[' -n '' ']'
+ return
+ return
+ '[' -n '' ']'
+ exit 1
```

The offending function is the following
```
# Run the named hook, either by calling the function with that name or
# by evaluating the variable with that name. This allows convenient
# setting of hooks both from Nix expressions (as attributes /
# environment variables) and from shell scripts (as functions). If you
# want to allow multiple hooks, use runHook instead.
_callImplicitHook() {
    local def="$1"
    local hookName="$2"
    if declare -F "$hookName" > /dev/null ; then
        "$hookName"
    elif type -p "$hookName" > /dev/null ; then
        source "$hookName"
    elif [ -n "${!hookName:-}" ]; then
        eval "${!hookName}"
    else
        return "$def"
    fi
    # `_eval` expects hook to need nounset disable and leave it
    # disabled anyways, so Ok to to delegate. The alternative of a
    # return trap is no good because it would affect nested returns.
}
```

since osh has errexit set by default and type fails, I think that is the problem here.
I'm going to try replacing the `> /dev/null` with `|| false`

happysalada commented:
That did not work. on strict errexit, adding `|| false` still generated the exit code 1 and made the thing fail.
I guess that the `type  -p` can't be used here. I'm trying `test -e` which is to test if a file exists (I thought it was to test if a variable exists).

happysalada committed:
setup.sh: replace type -p with test -e

type -p will exit 1 on failure.
Test makes the intent clearer here.

happysalada commented:
That was a superficial error hiding the following error
```
mapOffset relHostOffset
+ local -r pkg='/nix/store/y8j86z5r2a604ir8hckk30fvbgy5frr4-xz-5.2.5-bin'
+ local -r inputOffset=relHostOffset
+ declare -a envHostHostHooks= envHostTargetHooks=
+ local -r hostOffset=-1
+ (( "$inputOffset" <= 0 ))
          if (( "$inputOffset" <= 0 )); then
                ^
/nix/store/9zrxgrc25zb33pj1nhkga60wajzjmbj3-bootstrap-stage3-stdenv-darwin/setup:402: fatal: Invalid integer constant 'relHostOffset'
```

I have to say I'm a little puzzled at the mapOffset function being called with arguments that are variables that are not expanded. Here is the map offset function
```
    # The current package's host and target offset together
    # provide a <=-preserving homomorphism from the relative
    # offsets to current offset
    local mapOffsetResult
    function mapOffset() {
        local -r inputOffset="$1"
        if (( "$inputOffset" <= 0 )); then
            local -r outputOffset=$(( inputOffset + hostOffset ))
        else
            local -r outputOffset=$(( inputOffset - 1 + targetOffset ))
        fi
        mapOffsetResult="$outputOffset"
    }
```
and it's being called like
```
local relHostOffset
...
        mapOffset relHostOffset
```

I don't understand why it's not `mapOffset "$relHostOffset"`
I'm going to give it a try

happysalada commented:
I now get a different error, so I guess it was the right call
Here is the next error
```
[' -e '/nix/store/nr1ii2ss7fdp326ypqdlmndc3bq49mc4-xz-5.2.5-bin' ']'
+ local hostOffsetNext=-2
+ [[ "${allPlatOffsets[*]}" = *"$hostOffsetNext"*  ]]
+ local mapOffsetResult=
+ local relHostOffset=
+ continue
+ local files=propagatedBuildDepFiles
+ eval 'pkgsBuildHost+=("$pkg")'
+ mapOffset -1
+ local files=propagatedHostDepFiles
+ mapOffset 0
+ local -r inputOffset=0
+ (( "$inputOffset" <= 0 ))
+ local -r outputOffset=-1
+ mapOffsetResult=-1
+ local hostOffsetNext=-1
+ [[ "${allPlatOffsets[*]}" = *"$hostOffsetNext"*  ]]
+ pkgsBuildHost+=( '/nix/store/nr1ii2ss7fdp326ypqdlmndc3bq49mc4-xz-5.2.5-bin' )
+ '[' -e '/nix/store/nr1ii2ss7fdp326ypqdlmndc3bq49mc4-xz-5.2.5-bin' ']'
+ local mapOffsetResult=
+ local relHostOffset=
+ local files=propagatedBuildDepFiles
+ local -r inputOffset=-1
+ (( "$inputOffset" <= 0 ))
+ local -r outputOffset=-2
+ mapOffsetResult=-2
+ local hostOffsetNext=-2
+ [[ "${allPlatOffsets[*]}" = *"$hostOffsetNext"*  ]]
+ continue
+ local files=propagatedHostDepFiles
+ mapOffset 0
+ local relTargetOffset=
+ (( "$relHostOffset" <= "$relTargetOffset" ))
+ continue
+ (( "$relHostOffset" <= "$relTargetOffset" ))
+ mapOffset -1
+ local -r inputOffset=-1
+ (( "$inputOffset" <= 0 ))
+ local -r outputOffset=-2
+ mapOffsetResult=-2
+ local hostOffsetNext=-2
+ local -r inputOffset=0
+ (( "$inputOffset" <= 0 ))
+ local -r outputOffset=-1
+ mapOffsetResult=-1
+ local fileRef='propagatedHostDepFiles[0]'
+ local file=propagated-host-host-deps
+ unset -v fileRef
+ mapOffset 0
+ local -r inputOffset=0
+ (( "$inputOffset" <= 0 ))
+ local -r outputOffset=-1
+ local hostOffsetNext=-1
+ [[ "${allPlatOffsets[*]}" = *"$hostOffsetNext"*  ]]
+ local relTargetOffset=
+ (( "$relHostOffset" <= "$relTargetOffset" ))
+ continue
+ (( "$relHostOffset" <= "$relTargetOffset" ))
+ local fileRef='propagatedHostDepFiles[0]'
+ local file=propagated-host-host-deps
+ unset -v fileRef
+ mapOffsetResult=-1
+ mapOffset 0
+ local targetOffsetNext=-1
+ [[ "${allPlatOffsets[*]}" = *"$hostOffsetNext"*  ]]
+ local -r inputOffset=0
+ [[ "${allPlatOffsets[*]}" = *"$targetOffsetNext"* ]]
+ [[ -f "$pkg/nix-support/$file" ]]
+ continue
+ (( "$relHostOffset" <= "$relTargetOffset" ))
+ local fileRef='propagatedHostDepFiles[1]'
+ local file=propagated-build-inputs
+ continue
+ local files=propagatedHostDepFiles
+ mapOffset 0
+ local -r inputOffset=0
+ (( "$inputOffset" <= 0 ))
+ (( "$inputOffset" <= 0 ))
+ local -r outputOffset=-1
+ mapOffsetResult=-1
+ local targetOffsetNext=-1
+ [[ "${allPlatOffsets[*]}" = *"$targetOffsetNext"* ]]
+ [[ -f "$pkg/nix-support/$file" ]]
+ local -r outputOffset=-1
+ continue
+ unset -v fileRef
+ mapOffset 1
+ local -r inputOffset=1
+ (( "$inputOffset" <= 0 ))
+ local -r outputOffset=0
+ mapOffsetResult=0
+ local targetOffsetNext=0
+ mapOffsetResult=-1
+ local hostOffsetNext=-1
+ [[ "${allPlatOffsets[*]}" = *"$hostOffsetNext"*  ]]
+ local relTargetOffset=
+ (( "$relHostOffset" <= "$relTargetOffset" ))
+ continue
+ (( "$relHostOffset" <= "$relTargetOffset" ))
+ local fileRef='propagatedHostDepFiles[0]'
+ local file=propagated-host-host-deps
+ (( "$relHostOffset" <= "$relTargetOffset" ))
+ local fileRef='propagatedHostDepFiles[1]'
+ local file=propagated-build-inputs
+ unset -v fileRef
+ mapOffset 1
+ local -r inputOffset=1
+ (( "$inputOffset" <= 0 ))
+ findInputs '/nix/store/nr1ii2ss7fdp326ypqdlmndc3bq49mc4-xz-5.2.5-bin' -1 0
+ local -r pkg='/nix/store/nr1ii2ss7fdp326ypqdlmndc3bq49mc4-xz-5.2.5-bin'
+ local -r hostOffset=-1
+ local -r targetOffset=0
+ (( "$hostOffset" <= "$targetOffset" ))
+ local varVar=pkgBuildAccumVars
+ local varRef='pkgBuildAccumVars[1]'
+ [[ "${allPlatOffsets[*]}" = *"$targetOffsetNext"* ]]
+ [[ -f "$pkg/nix-support/$file" ]]
+ continue
+ unset -v fileRef
+ local -r outputOffset=0
+ mapOffsetResult=0
+ local var=pkgsBuildHost
+ unset -v varVar varRef
+ local files=propagatedTargetDepFiles
+ mapOffset 1
+ local -r inputOffset=1
+ (( "$inputOffset" <= 0 ))
+ local -r outputOffset=0
+ mapOffset 0
+ local -r inputOffset=0
+ (( "$inputOffset" <= 0 ))
+ local -r outputOffset=-1
+ local targetOffsetNext=0
+ [[ "${allPlatOffsets[*]}" = *"$targetOffsetNext"* ]]
+ [[ -f "$pkg/nix-support/$file" ]]
+ continue
+ local files=propagatedTargetDepFiles
+ local varSlice='pkgsBuildHost[*]'
+ unset -v varSlice
+ eval 'pkgsBuildHost+=("$pkg")'
+ mapOffsetResult=-1
+ mapOffsetResult=0
+ local hostOffsetNext=0
+ [[ "${allPlatOffsets[*]}" = *"$hostOffsetNext"*  ]]
+ local relTargetOffset=
+ (( "$relHostOffset" <= "$relTargetOffset" ))
+ continue
+ (( "$relHostOffset" <= "$relTargetOffset" ))
+ continue
+ (( "$relHostOffset" <= "$relTargetOffset" ))
+ local targetOffsetNext=-1
+ [[ "${allPlatOffsets[*]}" = *"$targetOffsetNext"* ]]
+ [[ -f "$pkg/nix-support/$file" ]]
+ continue
+ (( "$relHostOffset" <= "$relTargetOffset" ))
+ local fileRef='propagatedHostDepFiles[1]'
+ local file=propagated-build-inputs
+ unset -v fileRef
+ mapOffset 1
+ local -r inputOffset=1
+ (( "$inputOffset" <= 0 ))
+ local -r outputOffset=0
+ mapOffsetResult=0
+ local hostOffsetNext=0
+ [[ "${allPlatOffsets[*]}" = *"$hostOffsetNext"*  ]]
+ local relTargetOffset=
+ pkgsBuildHost+=( '/nix/store/nr1ii2ss7fdp326ypqdlmndc3bq49mc4-xz-5.2.5-bin' )
+ '[' -e '/nix/store/nr1ii2ss7fdp326ypqdlmndc3bq49mc4-xz-5.2.5-bin' ']'
+ local mapOffsetResult=
+ local relHostOffset=
+ local files=propagatedBuildDepFiles
+ mapOffset -1
+ (( "$relHostOffset" <= "$relTargetOffset" ))
+ mapOffset 1
+ local fileRef='propagatedTargetDepFiles[0]'
+ local file=propagated-target-target-deps
+ unset -v fileRef
+ local -r inputOffset=1
+ (( "$inputOffset" <= 0 ))
+ continue
+ (( "$relHostOffset" <= "$relTargetOffset" ))
+ continue
+ local -r inputOffset=-1
+ (( "$inputOffset" <= 0 ))
+ local -r outputOffset=0
+ local -r outputOffset=-2
+ mapOffset 1
+ local -r inputOffset=1
+ (( "$inputOffset" <= 0 ))
+ mapOffsetResult=0
+ local targetOffsetNext=0
+ [[ "${allPlatOffsets[*]}" = *"$targetOffsetNext"* ]]
+ [[ -f "$pkg/nix-support/$file" ]]
+ continue
+ (( "$relHostOffset" <= "$relTargetOffset" ))
+ local fileRef='propagatedTargetDepFiles[0]'
+ local file=propagated-target-target-deps
+ mapOffsetResult=-2
+ local hostOffsetNext=-2
+ [[ "${allPlatOffsets[*]}" = *"$hostOffsetNext"*  ]]
+ continue
+ local files=propagatedHostDepFiles
+ local -r outputOffset=0
+ mapOffsetResult=0
+ local targetOffsetNext=0
+ [[ "${allPlatOffsets[*]}" = *"$targetOffsetNext"* ]]
+ [[ -f "$pkg/nix-support/$file" ]]
+ continue
+ findInputs '/nix/store/mjjy30kxz775bhhi6j9phw81qh6dsbrf-move-docs.sh' -1 0
+ local files=propagatedTargetDepFiles
+ mapOffset 1
+ local -r inputOffset=1
+ (( "$inputOffset" <= 0 ))
+ local -r outputOffset=0
+ mapOffsetResult=0
+ local hostOffsetNext=0
+ unset -v fileRef
+ mapOffset 1
+ local -r inputOffset=1
+ (( "$inputOffset" <= 0 ))
+ local -r outputOffset=0
+ mapOffsetResult=0
+ local targetOffsetNext=0
+ mapOffset 0
+ local -r inputOffset=0
+ (( "$inputOffset" <= 0 ))
+ local -r outputOffset=-1
+ mapOffsetResult=-1
+ local hostOffsetNext=-1
+ [[ "${allPlatOffsets[*]}" = *"$hostOffsetNext"*  ]]
+ local relTargetOffset=
+ local -r pkg='/nix/store/mjjy30kxz775bhhi6j9phw81qh6dsbrf-move-docs.sh'
+ [[ "${allPlatOffsets[*]}" = *"$hostOffsetNext"*  ]]
+ [[ "${allPlatOffsets[*]}" = *"$targetOffsetNext"* ]]
+ (( "$relHostOffset" <= "$relTargetOffset" ))
+ local -r hostOffset=-1
+ local -r targetOffset=0
+ (( "$hostOffset" <= "$targetOffset" ))
+ local varVar=pkgBuildAccumVars
+ local relTargetOffset=
+ (( "$relHostOffset" <= "$relTargetOffset" ))
+ continue
+ (( "$relHostOffset" <= "$relTargetOffset" ))
+ continue
+ [[ -f "$pkg/nix-support/$file" ]]
+ continue
+ findInputs '/nix/store/mjjy30kxz775bhhi6j9phw81qh6dsbrf-move-docs.sh' -1 0
+ continue
+ (( "$relHostOffset" <= "$relTargetOffset" ))
+ local fileRef='propagatedHostDepFiles[0]'
+ local file=propagated-host-host-deps
+ local varRef='pkgBuildAccumVars[1]'
+ local var=pkgsBuildHost
+ (( "$relHostOffset" <= "$relTargetOffset" ))
+ local -r pkg='/nix/store/mjjy30kxz775bhhi6j9phw81qh6dsbrf-move-docs.sh'
+ unset -v fileRef
+ unset -v varVar varRef
+ local fileRef='propagatedTargetDepFiles[0]'
+ local file=propagated-target-target-deps
+ local -r hostOffset=-1
+ local -r targetOffset=0
+ mapOffset 0
+ local -r inputOffset=0
+ local varSlice='pkgsBuildHost[*]'
      case "${!varSlice-}" in
           ^
/nix/store/4sc142mdwjahf3qg06a0zs8rcvh9rppd-bootstrap-stage3-stdenv-darwin/setup:384: fatal: This word should yield a string, but it contains an array
```

matthewbauer commented on pkgs/stdenv/generic/setup.sh:
```diff
@@ -341,22 +342,22 @@ declare -a allPlatOffsets=(-1 0 1)
 # implements.
 findInputs() {
     local -r pkg="$1"
-    local -ri hostOffset="$2"
```
Does shellcheck say to not do this? It seems reasonable to set -i:

> The variable is to be treated as an integer; 
>  arithmetic evaluation is performed when the 
>  variable is assigned a value.

https://ss64.com/bash/declare.html

happysalada replied:
This is the more controversial change. -i does not work at the moment with osh.
Apparently, -i is basically a no-op in bash. The recommended way to do arithmetic is to use `((...))` where the -i does not make a difference.
There was a longer discussion though between the creator of osh and zimbatm
oilshell/oil#864
(I tried to link it in the commit message)

the -i does not come from shellcheck and could be made to work with osh, it would be a matter of making a PR upstream. At the moment it does not, so I just removed it.
(perhaps the commit title could have been clearer)

happysalada replied:
Here is more relevant things around this at http://mywiki.wooledge.org/BashGuide/Parameters
Specifically
```
However, in practice the use of declare -i is exceedingly rare. In large part, this is because it creates behavior that can be surprising to anyone trying to maintain the script, who misses the declare statement. Most experienced shell scripters prefer to use explicit arithmetic commands (with ((...)) or let) when they want to perform arithmetic. 
```

zimbatm replied:
Abandon all hopes of types, it's bash :D 

If I remember correctly, `-i` changes the parser on the fly. It also doesn't catch errors so it's not super useful. Eg: `declare -ri foo=bar && echo $foo` will output `0`

matthewbauer commented on pkgs/stdenv/generic/setup.sh:
```diff
@@ -995,10 +997,10 @@ configurePhase() {
-        # shellcheck disable=SC2086
-        local flagsArray=(
-            $configureFlags ${configureFlagsArray+"${configureFlagsArray[@]}"}
+        IFS=" " read -r -a configureFlagsTemp <<< "$configureFlags"
```
I'm not super familiar with this, but this looks like it works in the way we want, while avoiding the shellcheck warning.

/cc @ericson2314

matthewbauer replied:
This might be one of those things that was only an issue pre-Bash4?

happysalada replied:
This is coming from https://github.com/koalaman/shellcheck/wiki/SC2206
it's not so pretty, but it seems to work.

happysalada replied:
Additionally reading up on arrays, I found this, which recommends the same
http://mywiki.wooledge.org/BashGuide/Arrays

matthewbauer commented:
:+1: on this! I think all of these changes to the shell (or bashish) scripts are reasonable on their own. Any stats on comparisons to bash would be interesting, though.

That varSlice part can probably be rewritten based on the todo. It should just return 0 if $pkg is in the $var map.

happysalada commented:
@matthewbauer thank you for the explanation. I also found a reference commit [here](https://github.com/NixOS/nixpkgs/commit/81194eef45e2c03018687be60c2c695a1729df36#diff-7c7ca80928136cfc73a02d5b28350bd900e331d6d304857053ffc9f7beaad576)

I'm guessing it could be rewritten as 
```
    case "${var[@]}" in
        *" $pkg "*) return 0 ;;
    esac
```
testing this right now

happysalada committed:
shellcheck disable=SC2154

https://github.com/koalaman/shellcheck/wiki/SC2154

happysalada committed:
shellcheck disable=SC2148

https://github.com/koalaman/shellcheck/wiki/SC2148

happysalada committed:
setup.sh: shellcheck disable=SC1090

https://github.com/koalaman/shellcheck/wiki/SC1090

happysalada committed:
setup.sh: shellcheck disable=SC2123

https://github.com/koalaman/shellcheck/wiki/SC2123

happysalada committed:
setup.sh: fatal: Invalid integer constant 'relHostOffset'

dereference variables passed to mapOffset

happysalada committed:
setup.sh: add quotes where semantically valid

happysalada commented:
I had the following error
```
     case "${var[@]}" in
            ^~
/nix/store/0ywdaif5l83acmzx5d1ysg8miiyy07p7-bootstrap-stage3-stdenv-darwin/setup:382: fatal: Can't index string with @
```

happysalada commented:
I've solved it with the following code
```
    # check if $pkgs is in the array referenced by var
    case "${var}[@]" in
        *" $pkg "*) return 0 ;;
    esac
```

the next error seems to be different, so I would think it's a correct solution.
Although I'm a bit mixed on this one, I see also several statements like
```
[[ "${allPlatOffsets[*]}" = *"$targetOffsetNext"* ]]
```
Perhaps that is a better way of checking inclusion.
I'm going to research a little more on arrays before moving on to the next one


Here is the next error
```
local varName=_XDG_DATA_DIRS
+ export _PATH='/nix/store/85jd7nxhp2w2x64xf21iqx61p8nyh5pi-xz-5.2.5-bin/bin'
+ [[ "$hostOffset" -le -1 ]]
+ addToSearchPath _XDG_DATA_DIRS '/nix/store/85jd7nxhp2w2x64xf21iqx61p8nyh5pi-xz-5.2.5-bin/share'
+ [[ -d "$dir" && "${!varName:+${delimiter}${!varName}${delimiter}}" \ ...
+ export _PATH='/nix/store/85jd7nxhp2w2x64xf21iqx61p8nyh5pi-xz-5.2.5-bin/bin'
+ local dir='/nix/store/85jd7nxhp2w2x64xf21iqx61p8nyh5pi-xz-5.2.5-bin/share'
+ [[ -d "$dir" && "${!varName:+${delimiter}${!varName}${delimiter}}" \ ...
+ [[ "$hostOffset" -eq 0 && -d "$pkg/bin" ]]
+ local delimiter=':'
+ local varName=_PATH
+ local dir='/nix/store/85jd7nxhp2w2x64xf21iqx61p8nyh5pi-xz-5.2.5-bin/bin'
+ [[ -f "$pkg/nix-support/setup-hook" ]]
+ [[ "$hostOffset" -le -1 ]]
+ addToSearchPathWithCustomDelimiter ':' _XDG_DATA_DIRS '/nix/store/85jd7nxhp2w2x64xf21iqx61p8nyh5pi-xz-5.2.5-bin/share'
+ local delimiter=':'
+ local varName=_XDG_DATA_DIRS
+ local dir='/nix/store/85jd7nxhp2w2x64xf21iqx61p8nyh5pi-xz-5.2.5-bin/share'
+ addToSearchPath _XDG_DATA_DIRS '/nix/store/85jd7nxhp2w2x64xf21iqx61p8nyh5pi-xz-5.2.5-bin/share'
+ addToSearchPathWithCustomDelimiter ':' _XDG_DATA_DIRS '/nix/store/85jd7nxhp2w2x64xf21iqx61p8nyh5pi-xz-5.2.5-bin/share'
+ activatePackage '/nix/store/mjjy30kxz775bhhi6j9phw81qh6dsbrf-move-docs.sh' -1 0
+ local pkg='/nix/store/mjjy30kxz775bhhi6j9phw81qh6dsbrf-move-docs.sh'
+ local -r hostOffset=-1
+ local delimiter=':'
+ local -r targetOffset=0
+ [[ -d "$dir" && "${!varName:+${delimiter}${!varName}${delimiter}}" \ ...
+ [[ "$hostOffset" -eq 0 && -d "$pkg/bin" ]]
+ [[ -f "$pkg/nix-support/setup-hook" ]]
+ (( "$hostOffset" <= "$targetOffset" ))
+ '[' -f '/nix/store/mjjy30kxz775bhhi6j9phw81qh6dsbrf-move-docs.sh' ']'
+ source '/nix/store/mjjy30kxz775bhhi6j9phw81qh6dsbrf-move-docs.sh'
+ [[ -d "$dir" && "${!varName:+${delimiter}${!varName}${delimiter}}" \ ...
+ export _PATH='/nix/store/85jd7nxhp2w2x64xf21iqx61p8nyh5pi-xz-5.2.5-bin/bin'
+ [[ "$hostOffset" -le -1 ]]
+ activatePackage '/nix/store/mjjy30kxz775bhhi6j9phw81qh6dsbrf-move-docs.sh' -1 0
+ local pkg='/nix/store/mjjy30kxz775bhhi6j9phw81qh6dsbrf-move-docs.sh'
+ local varName=_XDG_DATA_DIRS
  preFixupHooks+=(_moveToShare)
  ^~~~~~~~~~~~~~~
/nix/store/mjjy30kxz775bhhi6j9phw81qh6dsbrf-move-docs.sh:5: fatal: Undefined variable 'preFixupHooks'
```

happysalada committed:
setup.sh: missing white space

happysalada committed:
setup.sh: shellcheck disable=SC2034

https://github.com/koalaman/shellcheck/wiki/SC2034

happysalada committed:
move-docs.sh: shellcheck 2166
https://github.com/koalaman/shellcheck/wiki/SC2166

happysalada committed:
move-docs.sh: quote where semanticaly valid

happysalada commented:
I'm trying the following
```
declare -a preFixHooks
```
adding this early in setup.sh to see if it works

I'm also trying checking for array inclusion with 
```
    # check if $pkgs is in the array referenced by var
    if [[ "${var}[*]" = *" $pkg "* ]]; then
        return 0
    fi
```
it seems a little more readable, but I'm no bash expert.

happysalada commented:
Here is interesting information against `set  -e`
http://mywiki.wooledge.org/BashFAQ/105

happysalada committed:
autopatchelf.sh: add link about grep -q

happysalada committed:
setup.sh: shellcheck disable=SC1091

https://github.com/koalaman/shellcheck/wiki/SC1091

happysalada committed:
setup.sh: fatal: This word should yield a string, but it contains an array

remove implicit array comparison in case

happysalada commented:
Next error is
```
source '/nix/store/kxw6q8v6isaqjm702d71n2421cxamq68-make-symlinks-relative.sh'
+ '[' -f '/nix/store/ajiadjpmg3mhqywbb1hlcaly9bajki55-xz-5.2.5-bin' ']'
  fixupOutputHooks+=(_makeSymlinksRelative)
  ^~~~~~~~~~~~~~~~~~
/nix/store/kxw6q8v6isaqjm702d71n2421cxamq68-make-symlinks-relative.sh:1: fatal: Undefined variable 'fixupOutputHooks'
```

Same idea, define it early on and see the next error.

happysalada commented:
Here is the next error
```
+ eval '_callImplicitHook 0 failureHook'
  export NIX_CFLAGS_COMPILE+=" -frandom-seed=$(
         ^~~~~~~~~~~~~~~~~~~~
/nix/store/pr6nzbscdpg94yvr151lrif2kg0csx7b-reproducible-builds.sh:5: fatal: += not allowed in assignment
```

happysalada committed:
setup.sh: fatal: Undefined variable

define hook arrays before assigning to them

happysalada commented:
I got to a new error, so I that means (I guess that the previous fix was alright)
Here is the new error
```
+ local visitedPkgs=''
                      ${pkgsBuildBuild+"${pkgsBuildBuild[@]}"} \
                      ^~
/nix/store/n9b1b7fj96y9750drm89zvs7m0pvq6sg-bootstrap-stage3-stdenv-darwin/setup:583: fatal: Array 'pkgsBuildBuild' can't be referred to as a scalar (without @ or *)
```

happysalada committed:
reproducible-builds.sh: fatal: += not allowed in assignment

take in account undefined case
assign and export in separate step
https://github.com/koalaman/shellcheck/wiki/SC2155

zimbatm commented on pkgs/build-support/setup-hooks/move-docs.sh:
```diff
@@ -6,16 +6,16 @@ preFixupHooks+=(_moveToShare)
 _moveToShare() {
     forceShare=${forceShare:=man doc info}
-    if [ -z "$forceShare" -o -z "$out" ]; then return; fi
+    if [ -z "$forceShare" ] || [ -z "$out" ]; then return; fi
```
```suggestion
    if [[ -z "$forceShare" || -z "$out" ]]; then return; fi
```
In general, I think we should use the `[[ ]]` bash tests everywhere. They allow that kind of simplified construct, and are also less vulnerable to unescaped variables.

happysalada replied:
I agree, I just followed the shellcheck to the letter here because I wanted this PR to be the least controversial as possible.
Let me change this

zimbatm commented on pkgs/stdenv/generic/setup.sh:
```diff
@@ -1,11 +1,12 @@
+# shellcheck disable=SC2148
```
It would be good to add an explanation attached to the disable. That way, future readers can know why without digging in the git history.

happysalada replied:
You mean a line under referencing the shellcheck for example?

SuperSandro2000 replied:
```suggestion
# shellcheck shell=bash
```

happysalada replied:
I've just added a commit that adds a link for the shellcheck disable in setup.sh
I don't want to go overboard on this yet, as I'm afraid this is going to get harder to merge.
I personally like having a link, it's much more outsider friendly.

zimbatm commented on pkgs/stdenv/generic/setup.sh:
```diff
@@ -62,7 +63,8 @@ _callImplicitHook() {
     if declare -F "$hookName" > /dev/null; then
         "$hookName"
-    elif type -p "$hookName" > /dev/null; then
+    elif test -e "$hookName" ; then
```
this is not the same thing:
```
$ type -p dirname && echo OK || echo FAIL
/run/current-system/sw/bin/dirname
OK
$ test -e dirname && echo OK || echo FAIL
FAIL
```
I'm only looking at the diff so maybe there is an explanation in the git history.

happysalada replied:
no you're right. I only thought about a file case. 
My problem was that type -p return a non-zero exit code on failure.
The reason I thought about a file only is that inside the conditional is a source. You can't source a dir right?

zimbatm commented on pkgs/stdenv/generic/setup.sh:
```diff
@@ -116,13 +118,14 @@ exitHandler() {
         times > "$NIX_BUILD_TOP/.times"
-        local -a times=($(cat "$NIX_BUILD_TOP/.times"))
+        local -a buildTimesArray
+        IFS=" " read -r -a buildTimesArray <<< "$(cat "$NIX_BUILD_TOP/.times")"
```
```suggestion
        IFS=" " read -r -a buildTimesArray < "$NIX_BUILD_TOP/.times"
```
`cat` is generally superfluous and can be replaced with `<`

happysalada replied:
It sounds right, let me test this.

zimbatm commented on pkgs/stdenv/generic/setup.sh:
```diff
@@ -1053,23 +1065,26 @@ checkPhase() {
         #TODO(@oxij): should flagsArray influence make -n?
         if make -n ${makefile:+-f $makefile} check >/dev/null 2>&1; then
-            checkTarget=check
+            checkTarget="check"
```
the RHS of variables don't need to be quoted

happysalada replied:
https://github.com/koalaman/shellcheck/wiki/SC2209

zimbatm reviewed (commented):
Good stuff. It's a good idea to exercise the stdlib with other implementations of bash.

happysalada committed:
prefer [[ over [ for multiple conditionals

Co-authored-by: Jonas Chevalier <zimbatm@zimbatm.com>

happysalada commented:
The next error on the list is
```
 if [ -z "${NIX_SSL_CERT_FILE:-}" ] && [ "${IN_NIX_SHELL:-}" != "impure" ]; then
                                                                           ^
/nix/store/cp0fim6w0wzp4cm9vfrsg52dwinihjjg-bootstrap-stage3-stdenv-darwin/setup:676: fatal: strict_errexit only allows simple commands (got command.Sentence). Hint: use 'run'.
```

I think it calls for a [[ instead of [, I'm going to try this

happysalada commented:
@zimbatm Thanks a lot for having a look at this! ❤️ 

I just have one question.
Once perl builds, what is the next step?

happysalada committed:
setup.sh: arithmetic conditional style

used "recommended" arithmetic conditional style

happysalada committed:
setup.sh: remove extraneous cat cmd

happysalada committed:
setup.sh: use [[ for combined conditionals

happysalada committed:
setup.sh: remove bash empty array old hack

SuperSandro2000 commented on pkgs/build-support/setup-hooks/move-docs.sh:
```diff
@@ -6,18 +6,17 @@ preFixupHooks+=(_moveToShare)
 _moveToShare() {
     forceShare=${forceShare:=man doc info}
-    if [ -z "$forceShare" -o -z "$out" ]; then return; fi
+    if [[ -z "$forceShare" || -z "$out" ]]; then return; fi
```
```suggestion
    if [[ -z $forceShare || -z $out ]]; then return; fi
```

happysalada replied:
I would tend to quote whenever possible to offer a consistent way of doing things. A little bit like the use of `const` and `var` in other languages. So that when the quotes are missing, it means the reader should really pay attention.

happysalada replied:
You're right, shellcheck doesn't complain about this one.

SuperSandro2000 commented on pkgs/stdenv/generic/builder.sh:
```diff
@@ -4,18 +4,20 @@ for i in $initialPath; do
 done
 
-mkdir $out
+mkdir "$out"
```
we don't need to quote every $out because it cannot contain spaces.

happysalada replied:
Agreed, however IMO it's better to try to quote in every case you can, you don't lose anything by doing it.
I feel this is more like a "style guide" than a real error.
Just like using [ over [[. Sometimes you can, but overall having a single style would make readability easier I think.

happysalada replied:
I'm happy to remove quotes where you think it's inappropriate. I have a different opinion, but I would like to focus on what we can all agree on. So if you let me know where you think I wrongly added quotes, I'll remove them.

zimbatm commented:
@happysalada I think the next step would be to benchmark a boostrap build and compare bash vs osh. To get a sense if it's worth pursuing that effort. Even if the stdenv doesn't get switched to osh, I think it's worth merging most of these fixes as they make the stdenv overall cleaner.

happysalada committed:
Update pkgs/stdenv/generic/setup.sh

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

happysalada commented:
Next error is a little bit harder
```
 role_post=_FOR_BUILD
      export NIX_LDFLAGS${role_post}+=" -lintl"
             ^~~~~~~~~~~
/nix/store/x4m59y2c6j0vxbvkvi9c5bqsx2mfsqgz-gettext-0.21/nix-support/setup-hook:87: fatal: Invalid variable name 'NIX_LDFLAGS_FOR_BUILD+= -lintl'
```
It seems this is missing an eval somewhere.

happysalada committed:
shellcheck: shell=bash

happysalada committed:
autoreconf.sh: define preConfigurePhases before assigning

happysalada commented:
The astute observer will have remarked upon missing double quotes in the previous suggestion.

Here is the next error
```
         export NIX_CFLAGS_COMPILE${role_post}+=" -isystem $1/include"
                 ^~~~~~~~~~~~~~~~~~
/nix/store/sl1hh8zcx496w0kf39mwdla63lp65zwm-bootstrap-stage0-clang-wrapper-7.1.0/nix-support/setup-hook:143: fatal: Invalid variable name 'NIX_CFLAGS_COMPILE+= -isystem /nix/store/s423q4a78w84dh9f0m4xsd2aq8pyhdl0-python3-minimal-3.8.9/include'
  source $stdenv/setup
```

happysalada committed:
setup.sh: separate command from combined conditional

happysalada committed:
setup-hook.sh: use eval for dynamic variable export

happysalada commented:
To everybody who has taken the time to review, thank you so much!

I would like to propose merging what I have so far. Since those are only "small" shell modifications, I'm hoping it won't be too controversial. I will continue afterwards, but I'm afraid in getting too far and not being able to merge.

Happy to discuss where you have strong feelings. Also I'm not sure who should be brought to have a look exactly, or how long this should be left open. @zimbatm if you have opinions or ideas, I'm interested. No worries if you are busy of course.

aszlig commented on pkgs/build-support/setup-hooks/multiple-outputs.sh:
```diff
@@ -75,6 +79,7 @@ _multioutConfig() {
         --localedir=${!outputLib}/share/locale \
         $configureFlags"
 
+    echo $configureFlags
```
Is this a leftover from debugging?

happysalada replied:
good catch!

happysalada replied:
I'm so happy it was the last commit and I don't have to do crazyness to edit the commit.
Thank you for spotting this!

Artturin replied:
rebase fixup isn't craziness https://git-rebase.io/#fixup

happysalada replied:
I didn't know about that, thanks! ❤️

happysalada commented:
Just for information, there is one error after that, where I'm not really sure what is happening
```
+ '[' -d '/nix/store/izmjbqg7a2ysam4v411n0wfwphhl7pa9-libtool-2.4.6-lib/include' ']'
+ '[' -d '/nix/store/izmjbqg7a2ysam4v411n0wfwphhl7pa9-libtool-2.4.6-lib/Library/Frameworks' ']'
  NIX_CC_USE_RESPONSE_FILE=0 "$CC" -std=c99 -O3 -o "expand-response-params" expand-response-params.c
                             ^
[ eval at line 1378 of /nix/store/dzlq0ss3p2c3nn9snbdm6agnp51sxc4h-bootstrap-stage3-stdenv-darwin/setup ]:1378: fatal: Exiting with status 2 (command in PID 74630)
```

we can leave that for the next PR I thing.

happysalada committed:
multiple-outputs.sh: remove grep from combined conditional

happysalada commented:
Another question I have is the default dependency on readline in oil.
The default version of oil includes interactiveness with a dependency on readline.
I was thinking of splitting oil into "oil-uninteractive" and oil that includes interactiveness.
Most people using oil will want interactiveness, so the default IMO should be to include it. Then in the bootstrap, having an additional dependency being a potential problem, it could use if we end up deciding upon it "oil-uninteractiv" or  "oil-raw", not sure what is the best naming for it.

SuperSandro2000 commented:
> I was thinking of splitting oil into "oil-uninteractive" and oil that includes interactiveness.

Why not follow bash and bashInteractive?

SuperSandro2000 reviewed (changes requested):
I don't think we need to link any of the shellcheck wikis as you can find all of them by simply throwing the shellcheck code into google.

SuperSandro2000 commented on pkgs/stdenv/generic/setup.sh:
```diff
@@ -843,9 +863,12 @@ _defaultUnpack() {
 
 unpackFile() {
+    local status
     curSrc="$1"
     header "unpacking source archive $curSrc" 3
-    if ! runOneHook unpackCmd "$curSrc"; then
+    runOneHook unpackCmd "$curSrc"
+    status=$?
+    if (( status != 0)); then
```
We don't need to capture the exit code. We do nothing with it.

happysalada replied:
So the problem here is that the conditional is a little complicated, osh doesn't like it.
I thought we would just run the command and check the status in a separate conditional.

happysalada replied:
Here is an additonal thought, the semantics here have changed a little. If the unpack fails, this will exit on errexit.
I'm thinking that is what we want.

SuperSandro2000 commented on pkgs/stdenv/generic/setup.sh:
```diff
@@ -530,15 +548,15 @@ _activatePkgs
 # doesn't have closures, so it's easier to just pass this in.
 _addToEnv() {
-    local -i depHostOffset depTargetOffset
+    local depHostOffset depTargetOffset
     local pkg
```
```suggestion
    local depHostOffset depTargetOffset pkg
```

SuperSandro2000 commented on pkgs/stdenv/generic/setup.sh:
```diff
@@ -497,19 +513,21 @@ activatePackage() {
 
 _activatePkgs() {
-    local -i hostOffset targetOffset
+    local hostOffset targetOffset
     local pkg
```
```suggestion
    local hostOffset targetOffset pkg
```

happysalada commented on pkgs/build-support/setup-hooks/autoreconf.sh:
```diff
@@ -1,4 +1,5 @@
-preConfigurePhases+=" autoreconfPhase"
+# shellcheck shell=bash
+preConfigurePhases="${preConfigurePhases:-} autoreconfPhase"
```
might not be defined.
This form makes sure if the variable is not defined, it will use an empty string instead

andychu replied:
FWIW I don't believe this should be affected by OSH but I can see why you would want to change it

happysalada commented on pkgs/stdenv/generic/setup.sh:
```diff
@@ -629,7 +647,7 @@ export NIX_INDENT_MAKE=1
 if [ -z "${NIX_BUILD_CORES:-}" ]; then
   NIX_BUILD_CORES="1"
-elif [ "$NIX_BUILD_CORES" -le 0 ]; then
+elif (( "$NIX_BUILD_CORES" <= 0 )); then
```
The "recommended" way to do arithmetic is with ((...)) or with $((...)).
The `-le` construct only works with single square brackets, with double you should be able to use `<=`.
I would like to stick to the "blessed" way, I'm happy to hear other opinions if people have any.

SuperSandro2000 replied:
``[[ 1 -le 2 ]]`` works in my bash 5.1.

happysalada replied:
I was sure I read that somewhere. Well, I stand corrected!
Can we agree on the `<=` being a little more "friendly" than the `-le` ?

happysalada replied:
Wow interesting!

happysalada replied:
the "recommended" way to do arithmetic is to use ((...))
I'm happy to revert that commit if you feel strongly about not using that syntax

happysalada commented:
> > I was thinking of splitting oil into "oil-uninteractive" and oil that includes interactiveness.
> 
> Why not follow bash and bashInteractive?

My thinking was that it forces users to make a decision to choose the interactive version (which should be what they want in 99% of the cases). I thought making interactive the default would make more sense. IMO the only time you really want non interactive is when you care about an additional dependency to readline (I can't think of anybody choosing this version except for a potential use in stdenv, I could be wrong though).

happysalada committed:
Update pkgs/stdenv/generic/setup.sh: group var declaration

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

happysalada committed:
Update pkgs/stdenv/generic/setup.sh: group var declaration

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

happysalada committed:
Update pkgs/stdenv/generic/setup.sh: use [[ instead of [

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

happysalada committed:
Update pkgs/stdenv/generic/setup.sh:use [[ instead of [

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

happysalada commented:
@SuperSandro2000 thanks a lot for taking the time to review ❤️

happysalada committed:
Update pkgs/build-support/setup-hooks/move-docs.sh: remove quote in conditional

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

happysalada commented:
I had some commits break building bash. After a couple of git bisect, I removed the commits.
I will re-work the problematic commits and include them in the next PR.
I plan on leaving this open until monday.

happysalada commented:
It seems the eval has not started yet. I'm going to change the branch to master to see if this will trigger a faster eval.
I don't intend to merge on master, I will just change it back once the eval is done.

happysalada commented:
eval passed, I'm merging this on staging.
I've kept all the other changes. I'm going to make a smaller PR next.
I'll keep notice of what can be controversial and will only try to submit things that should be ok.

andychu commented on pkgs/build-support/bintools-wrapper/setup-hook.sh:
```diff
@@ -14,7 +15,9 @@ bintoolsWrapper_addLDVars () {
 
     if [[ -d "$1/lib64" && ! -L "$1/lib64" ]]; then
-        export NIX_LDFLAGS${role_post}+=" -L$1/lib64"
+        varName=NIX_LDFLAGS${role_post}
```
I would add `local varName` so it doesn't "leak".  Also it seems like the style is `local var_name`, e.g. akin to `role_post` ?

It took me awhile to figure this code out but I get what it's doing now... it is indeed very dynamic.

I think `export "$var_name"` is sufficient?

andychu commented on pkgs/stdenv/generic/builder.sh:
```diff
@@ -4,18 +4,20 @@ for i in $initialPath; do
+    echo "defaultBuildInputs=\"$defaultBuildInputs\""
+    echo "$preHook"
+    cat "$setup"
+} > "$out/setup"
```
This is a great change!  One of my pet peeves :)

The `>>` style is less readable and actually causes a ton of `open()` and `close()` syscalls

andychu commented on pkgs/stdenv/generic/builder.sh:
```diff
@@ -4,18 +4,20 @@ for i in $initialPath; do
 # in stdenv.
 mkdir $out/nix-support
-if [ "$propagatedUserEnvPkgs" ]; then
+if [ -n "${propagatedUserEnvPkgs:-}" ]; then
```
A few comments:

1. FWIW this shouldn't be affected by OSH.   That is, strictly speaking, no change is necessary (I worry about people being confused based on the PR).
2. And it seems like the `:-` is a behavior change and not a style change?  Maybe that is intended.
3. Since you're using `[[` everywhere, you could also use it here.

happysalada replied:
Actually this is interesting.
I've been running it as `/bin/osh -x -O strict:all`
just to be able to debug and to get "potential" bugs early.
This variable was actually undefined in some cases and so the script could fail.
What you are saying is that I should not run it with strict:all and just try to see what comes out of it?

andychu commented on pkgs/stdenv/generic/setup.sh:
```diff
@@ -116,13 +118,14 @@ exitHandler() {
         times > "$NIX_BUILD_TOP/.times"
-        local -a times=($(cat "$NIX_BUILD_TOP/.times"))
+        local -a buildTimesArray
+        IFS=" " read -r -a buildTimesArray < "$NIX_BUILD_TOP/.times"
```
FWIW I thought the original was OK, and this shouldn't be affected by OSH?  I would be interested to learn otherwise.

happysalada replied:
This one was just a shell check

andychu commented on pkgs/stdenv/generic/setup.sh:
```diff
@@ -341,27 +351,24 @@ declare -a allPlatOffsets=(-1 0 1)
 
-    local varVar="${pkgAccumVarVars[$hostOffset + 1]}"
-    local varRef="$varVar[\$targetOffset - \$hostOffset]"
+    local varVar="${pkgAccumVarVars[$(( hostOffset + 1 ))]}"
```
BTW this can be `${vars[hostOffset + 1]}`, because whatever is within `${a[i]}` is parsed like arithemetic.  Maybe that is too subtle for some but I'm used to it.

(Also now I notice the naming style is sometimes `camelCase` and sometimes `role_host`)

I guess it is pretty confusing because below you there is some "lazy evaluation" with `${varVar}[  $(( expr )) ]`, which indeed does need `$(( ))`, because it's constructing a string to be evaluated later.  This code is very confusing :)

happysalada replied:
I don't like these references that much, but it seems that logic is in various places, I'm still trying to find a "better" solution, but I have to be able to understand the code holistically and I'm not there.

andychu commented on pkgs/stdenv/generic/setup.sh:
```diff
@@ -530,15 +538,14 @@ _activatePkgs
     for depHostOffset in "${allPlatOffsets[@]}"; do
-        local hookVar="${pkgHookVarVars[$depHostOffset + 1]}"
-        local pkgsVar="${pkgAccumVarVars[$depHostOffset + 1]}"
+        local hookVar="${pkgHookVarVars[$(( depHostOffset + 1 ))]}"
```
ditto here and throughout, $(( )) is optional in a bash/ksh array subscript

happysalada replied:
Thank you for this one, I'll keep this in mind!

andychu reviewed (commented):
(author of Oil here)

I realize this is already merged, but I just saw it and figured I would make some comments.  Feel free to ignore them :)

To be honest I find the commit description misleading, because it feels like 90% of the changes have nothing to do with making it run under OSH?  The changes I see that Oil probably won't run are:

- the `export FOO+=` one.  
- the `local -ri` changes

Other changes seem to have more to do with ShellCheck than OSH?  And just misc style changes like changing `[` to `[[` (which I generally agree with, but not really related to OSH).

ShellCheck and OSH are mostly complementary, but for my own edification and that of future OSH users/migrators, I would have liked to see the two groups of changes separated.  Personally I would be scared to make all these foundational changes to what looks like fairly fundamental code in one big bunch, but you know the code better than I !

happysalada commented:
@andychu Thanks again for your comment!
I've been running under `/bin/osh -x -O strict:all` and that was probably why some stuff where changed. I'm going to just try running it with `-O oil:basic` and see if it changes for the better. I guess trying to run it under strict is a little overoptimistic for now. I'll cc you in the next PR!

andychu commented:
Ah OK so `strict:all` turns on `errexit`, which can cause many changes.  (`oil:basic` also turns on `errexit`.)

It's not surprising that turning on `errexit` requires changes to shell code, which aren't strictly necessary for OSH.

If you're NOT running this script under bash with `set -e`, and are not *planning* to, then it could have caused you to change some stuff that isn't strictly necessary, like using `${x:-}` instead of `$x`.

But maybe you are planning to, and it's fine!  It's up to you either way, but I guess I would expect the PR to also set `-e` in bash if that's the case. 

I also wonder if we should turn off `errexit` in `strict:all` but keep it on in `oil:basic`.  I'm open to feedback on that.  Step 1 here is to just run `osh myscript.sh`, and we don't necessarily mention `strict:all`, although I think it's useful: https://github.com/oilshell/oil/wiki/Gradually-Upgrading-Shell-to-Oil

happysalada commented:
Thank you for sharing.
Just like you shared in the resource, I think the next step would be to not run it with any options at all and see what errors come out then. Perhaps try to fix those first.
I'll make it clearer on the next PR which fixes are for errexit and which are not, at least people have more information to judge if they want the fixes or not.

vcunat commented:
This seems to cause a weird regressions in some cases: https://github.com/NixOS/nixpkgs/pull/129506#issuecomment-877664572

I suspect it's cases where stdenv is used in some more complex way, e.g. the 64+32-bit for wine.  Anyway, if we can't pinpoint it rather soon, I'd suggest to revert for now until we can solve it, so other changes don't get blocked by this.

aszlig commented:
@vcunat: As mentioned in https://github.com/NixOS/nixpkgs/commit/bf99a819a16089a8df03cc06262f06b31dfb683c#r53315516, I'd suggest reverting that commit for now instead of the whole pull request.

vcunat commented:
For reference, this whole PR got reverted (for now).  It caused multiple unexpected regressions.  If you want to redo this, I suggest that:
- you pick some good commit from https://hydra.nixos.org/jobset/nixpkgs/trunk/evals and (re-)base the work on that;
- poke someone (probably me) to create a separate Hydra jobset for your branch (say, just x86_64-linux), so we can then better see and solve any regressions _before_ merging this again (Hydra can show diffs);
- ideally also get a bit more review of the code.

happysalada commented:
I'm going to try to separate this into logical parts so hopefully the simple fixes can be merged first.
I'll cc you on the first PR and includes the subsequent PRs in the description.
